core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2730,10 +2730,30 @@ object ZIO {
    */
   def reduceAllPar[R, R1 <: R, E, A](a: ZIO[R, E, A], as: Iterable[ZIO[R1, E, A]])(
     f: (A, A) => A
-  ): ZIO[R1, E, A] =
-    as.foldLeft[ZIO[R1, E, A]](a) { (l, r) =>
-      l.zipPar(r).map(f.tupled)
-    }
+  ): ZIO[R1, E, A] = {
+    def prepend(a: ZIO[R, E, A], as: Iterable[ZIO[R1, E, A]]): Iterable[ZIO[R1, E, A]] =
+      new Iterable[ZIO[R1, E, A]] {
+        override def iterator: Iterator[ZIO[R1, E, A]] = new Iterator[ZIO[R1, E, A]] {
+          private var started    = false
+          private val asIterator = as.iterator
+
+          override def hasNext: Boolean = !started || asIterator.hasNext
+
+          override def next(): ZIO[R1, E, A] =
+            if (!started) {
+              started = true
+              a
+            } else {
+              asIterator.next()
+            }
+        }
+      }
+
+    val all = prepend(a, as)
+    mergeAllPar(all)(Option.empty[A]) { (acc, elem) =>
+      Some(acc.fold(elem)(f(_, elem)))
+    }.map(_.get)
```

Comment:
Is it better to avoid unnecessary allocations and just deal with null?

```scala
mergeAllPar(all)(null) { (acc, elem) =>
  if (acc eq null) elem else f(acc, elem)
}
```

Reply:
It looks a lot like `mergeAllPar(as)(a)(f)` 🤔

Reply:
If we pass `a` as a second argument, then `f` should be of type `(ZIO[R, E, A], A) => ZIO[R, E, A]`. So, doesn't match.

Reply:
The allocation overhead is probably miniscule compared to the concurrency overhead, so I wouldn't bother with working with `null` here.

Reply:
Sure, I was just questioning the using of `Option`, is all.
```scala
a.map(mergeAllPar(as)(_)(f))
```
perhaps?

Reply:
Then `a` isn't run in parallel with `as`. :(

Reply:
Fair enough, that's too bad but hey the numbers are awesome!

Reply:
Yes, I think unfortunately we need this. We could use `zipWithPar` to execute the `a` in parallel but then we don't have a zero value for the merge operation.